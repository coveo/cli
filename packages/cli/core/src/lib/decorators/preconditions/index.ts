// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
