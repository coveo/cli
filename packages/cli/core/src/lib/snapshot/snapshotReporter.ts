import {
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportOperationModel,
  ResourceSnapshotsReportOperationResult,
  ResourceSnapshotType,
} from '@coveo/platform-client';
import {
  ReportViewerOperationName,
  ReportViewerResourceReportModel,
  SnapshotReportStatus,
} from './reportPreviewer/reportPreviewerDataModels';

type ResourceEntries = [string, ResourceSnapshotsReportOperationModel];

type NoopHandler = (this: void) => void;
type SnapshotReporterHandler = (this: SnapshotReporter) => void | Promise<void>;

type SnapshotReporterHandlers = {
  [K in SnapshotReportStatus]: SnapshotReporterHandler | NoopHandler;
};

const MissingVaultEntryResultCode = 'INVALID_PLACEHOLDER';
//#endregion

export type VaultEntryAttributes = {
  vaultEntryId: string;
  resourceName: string;
  resourceType: ResourceSnapshotType;
};

export class SnapshotReporter {
  private missingVaultEntriesSet: Set<VaultEntryAttributes> = new Set();

  private operationResultsParseFuseUsed: boolean = false;

  private consumeOperationResultsParseFuse() {
    if (this.operationResultsParseFuseUsed) {
      return false;
    } else {
      this.operationResultsParseFuseUsed = true;
      return true;
    }
  }

  public get missingVaultEntries() {
    return this.missingVaultEntriesSet.values();
  }
  public resourceInErrorCount: number = 0;
  public resourceInError: Map<Partial<ResourceSnapshotType>, Set<string>> =
    new Map();
  public constructor(public readonly report: ResourceSnapshotsReportModel) {}

  public static convertResourceEntriesToResourceSnapshotsReportOperationModel([
    name,
    operations,
  ]: ResourceEntries) {
    return {name, operations};
  }

  public static resourceHasAtLeastOneOperationFilter(
    operationsToDisplay: ReportViewerOperationName[]
  ) {
    return ([_, operations]: ResourceEntries) =>
      operationsToDisplay.reduce(
        (previous, current) => previous + operations[current],
        0
      ) > 0;
  }

  public getAffectedResourceTypes(): ResourceSnapshotType[] {
    const isResourceTypeReportAffected = (
      report: ResourceSnapshotsReportOperationModel
    ) => Object.values(report).some((operation) => operation > 0);
    const changedResourcesTypes: ResourceSnapshotType[] = [];
    for (const resourceType of Object.keys(this.report.resourceOperations)) {
      if (
        isResourceTypeReportAffected(
          this.report.resourceOperations[resourceType]
        )
      ) {
        changedResourcesTypes.push(resourceType as ResourceSnapshotType);
      }
    }
    return changedResourcesTypes;
  }

  public hasChangedResources() {
    const totalUnchanges =
      this.getOperationTypeTotalCount('resourcesUnchanged');
    return totalUnchanges !== this.report.resourcesProcessed;
  }

  public getOperationTypeTotalCount(
    type: keyof ResourceSnapshotsReportOperationModel
  ) {
    const count = Object.values(this.report.resourceOperations).reduce(
      (count: number, current: ResourceSnapshotsReportOperationModel) =>
        count + current[type],
      0
    );

    return count;
  }

  private reportHandlers: SnapshotReporterHandlers = {
    [SnapshotReportStatus.SUCCESS]: () => {},
    [SnapshotReportStatus.NO_CHANGES]: () => {},
    [SnapshotReportStatus.MISSING_VAULT_ENTRIES]: () => {},
    [SnapshotReportStatus.ERROR]: () => {},
  };

  public setReportHandler<T extends SnapshotReportStatus>(
    status: T,
    handler: SnapshotReporterHandlers[T]
  ): SnapshotReporter {
    this.reportHandlers[status] = handler;
    return this;
  }

  public async handleReport(): Promise<void> {
    const reportStatuses = this.getReportStatuses();
    await this.executeHandlers(reportStatuses.fixables);
    if (reportStatuses.errors.length > 0) {
      await this.executeHandlers(reportStatuses.errors);
      return;
    }
    if (reportStatuses.successes.length > 0) {
      await this.executeHandlers(reportStatuses.successes);
    }
  }

  private async executeHandlers(reportStatuses: SnapshotReportStatus[]) {
    for (const status of reportStatuses) {
      await this.callAndReset(status);
    }
  }

  private async callAndReset(handler: SnapshotReportStatus) {
    await this.reportHandlers[handler].apply(this);
    this.reportHandlers[handler] = () => {};
  }

  private getReportStatuses() {
    const statuses: Record<
      'fixables' | 'successes' | 'errors',
      SnapshotReportStatus[]
    > = {
      fixables: [],
      successes: [],
      errors: [],
    };

    this.parseResourcesOperationsResults();
    if (this.missingVaultEntriesSet.size > 0) {
      statuses.fixables.push(SnapshotReportStatus.MISSING_VAULT_ENTRIES);
    }

    if (!this.hasChangedResources()) {
      statuses.successes.push(SnapshotReportStatus.NO_CHANGES);
    }

    if (this.resourceInErrorCount > 0) {
      statuses.errors.push(SnapshotReportStatus.ERROR);
    } else {
      statuses.successes.push(SnapshotReportStatus.SUCCESS);
    }

    return statuses;
  }

  private parseResourcesOperationsResults(): void {
    if (!this.consumeOperationResultsParseFuse()) {
      return;
    }
    for (const [resourceType, resource] of Object.entries(
      this.report.resourceOperationResults
    )) {
      for (const [resourceName, errors] of Object.entries(resource)) {
        for (const err of errors) {
          this.parseResourceOperationResult(err, resourceName, resourceType);
        }
      }
    }
  }

  private parseResourceOperationResult(
    err: ResourceSnapshotsReportOperationResult,
    resourceName: string,
    resourceType: string
  ) {
    const missingEntry = SnapshotReporter.tryGetMissingVaultEntryName(err);
    if (missingEntry) {
      this.missingVaultEntriesSet.add({
        vaultEntryId: missingEntry,
        resourceName: resourceName,
        resourceType: resourceType as ResourceSnapshotType,
      });
    } else {
      // TODO: Fix PlatformClient to reflect proper typing.
      this.addResourceInError(resourceType as ResourceSnapshotType, err);
    }
  }

  private addResourceInError(
    resourceType: ResourceSnapshotType,
    err: ResourceSnapshotsReportOperationResult
  ) {
    this.resourceInErrorCount++;
    let errorSet = this.resourceInError.get(resourceType);
    if (!errorSet) {
      errorSet = new Set();
      this.resourceInError.set(resourceType, errorSet);
    }
    errorSet.add(`${err.resultCode}: ${err.message}`);
  }

  private static missingVaultEntryMatcher =
    /^The vault entry referenced by \{\{ VAULT\.(?<entryName>.*) \}\} could not be found in the vault\.$/;

  private static tryGetMissingVaultEntryName(
    err: ResourceSnapshotsReportOperationResult
  ): string | undefined {
    let message: string = '';
    if (typeof err !== 'string') {
      if (err.resultCode === MissingVaultEntryResultCode) {
        message = err.message;
      } else {
        return undefined;
      }
    } else {
      message = err;
    }
    const match = message.match(SnapshotReporter.missingVaultEntryMatcher);
    return match?.groups?.['entryName'];
  }

  public getChangedResources(
    operationsToDisplay: ReportViewerOperationName[]
  ): ReportViewerResourceReportModel[] {
    return Object.entries(this.report.resourceOperations)
      .filter(
        SnapshotReporter.resourceHasAtLeastOneOperationFilter(
          operationsToDisplay
        )
      )
      .map(
        SnapshotReporter.convertResourceEntriesToResourceSnapshotsReportOperationModel
      );
  }

  public get type() {
    const type = this.prettify(this.report.type);
    return type;
  }

  public get status() {
    const status = this.prettify(this.report.status);
    return status;
  }

  public get resultCode() {
    const status = this.prettify(this.report.resultCode);
    return status;
  }

  private prettify(str: string): string {
    const capitalized = (word: string) =>
      word.charAt(0) + word.slice(1).toLowerCase();

    const spaced = str.replace(/_/g, ' ');
    return spaced.split(' ').map(capitalized).join(' ');
  }
}
