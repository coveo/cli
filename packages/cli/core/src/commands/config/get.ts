import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
