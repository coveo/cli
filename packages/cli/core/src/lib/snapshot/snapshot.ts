import {
  ResourceSnapshotsModel,
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportResultCode,
  ResourceSnapshotsReportStatus,
  PlatformClient,
  ResourceSnapshotsReportType,
  SnapshotExportContentFormat,
  ApplyOptionsDeletionScope,
} from '@coveo/platform-client';
import retry from 'async-retry';
import {ReportViewer} from './reportPreviewer/reportPreviewer';
import {ensureFileSync, writeJsonSync} from 'fs-extra';
import {join} from 'path';
import {SnapshotReporter} from './snapshotReporter';
import {SnapshotOperationTimeoutError} from '../errors';
import {ExpandedPreviewer} from './expandedPreviewer/expandedPreviewer';
import {Project} from '../project/project';
import {SnapshotNoReportFoundError} from '../errors/snapshotErrors';
import {SnapshotReportStatus} from './reportPreviewer/reportPreviewerDataModels';

export type SnapshotReport = ResourceSnapshotsReportModel;

export interface WaitUntilDoneOptions {
  /**
   * The maximum number of seconds to wait before the commands exits with a timeout error.
   * A value of zero will prevent the command from timing out.
   */
  wait?: number; // in seconds
  /**
   * The interval between 2 consecutive polls.
   */
  waitInterval?: number; // in seconds
  /**
   * Callback to execute every time a request is being made to retrieve the snapshot data
   */
  onRetryCb?: (report: ResourceSnapshotsReportModel) => void;
}

export interface WaitUntilOperationDone extends WaitUntilDoneOptions {
  /**
   * The operation to wait for. If not specified, the method will wait for any operation to complete.
   */
  operationToWaitFor?: ResourceSnapshotsReportType;
}

export class Snapshot {
  public static defaultWaitOptions: Required<WaitUntilDoneOptions> = {
    waitInterval: 1,
    wait: 60,
    onRetryCb: (_report: ResourceSnapshotsReportModel) => {},
  };

  private static ongoingReportStatuses = [
    ResourceSnapshotsReportStatus.Pending,
    ResourceSnapshotsReportStatus.InProgress,
  ];

  public constructor(
    private model: ResourceSnapshotsModel,
    private client: PlatformClient
  ) {}

  public async validate(
    deleteMissingResources = false,
    options: WaitUntilDoneOptions = {}
  ): Promise<SnapshotReporter> {
    await this.snapshotClient.dryRun(this.id, {
      deleteMissingResources,
    });

    await this.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.DryRun,
      ...options,
    });

    return new SnapshotReporter(this.latestReport);
  }

  public async preview(
    projectToPreview: Project,
    deleteMissingResources = false,
    expandedPreview = true
  ) {
    const reporter = new SnapshotReporter(this.latestReport);
    await this.displayLightPreview(reporter);
    if (!expandedPreview) {
      return;
    }
    const onSuccess = () =>
      this.generateExpandedPreview(projectToPreview, deleteMissingResources);
    await reporter
      .setReportHandler(
        SnapshotReportStatus.SUCCESS,
        async function (this: SnapshotReporter) {
          await onSuccess();
        }
      )
      .handleReport();
  }

  public async apply(
    deleteMissingResources = false,
    options: WaitUntilDoneOptions = {}
  ) {
    await this.snapshotClient.apply(
      this.id,
      deleteMissingResources
        ? {
            deleteMissingResources,
            deletionScope: ApplyOptionsDeletionScope.OnlyTypesFromSnapshot,
          }
        : {deleteMissingResources}
    );

    await this.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.Apply,
      ...options,
    });

    return new SnapshotReporter(this.latestReport);
  }

  public async delete() {
    await this.client.resourceSnapshot.delete(this.model.id);
  }

  public download() {
    return this.client.resourceSnapshot.export(this.id, {
      contentFormat: SnapshotExportContentFormat.SplitPerType,
    });
  }

  public areResourcesInError() {
    return (
      this.latestReport.resultCode ===
      ResourceSnapshotsReportResultCode.ResourcesInError
    );
  }

  public saveDetailedReport(projectPath: string) {
    const pathToReport = join(
      projectPath,
      'snapshot-reports',
      `${this.latestReport.id}.json`
    );
    ensureFileSync(pathToReport);
    writeJsonSync(pathToReport, this.latestReport, {spaces: 2});
    return pathToReport;
  }

  public get latestReport() {
    const reports = this.model.reports;
    if (!Array.isArray(reports) || reports.length === 0) {
      throw new SnapshotNoReportFoundError(this);
    }
    return this.sortReportsByDate(reports)[0];
  }

  public get id() {
    return this.model.id;
  }

  public get targetId() {
    return this.model.targetId;
  }

  private get snapshotClient() {
    return this.client.resourceSnapshot;
  }

  private sortReportsByDate<T extends SnapshotReport>(report: T[]): T[] {
    return report.sort((a, b) => b.updatedDate - a.updatedDate);
  }

  private async displayLightPreview(reporter: SnapshotReporter) {
    const viewer = new ReportViewer(reporter);
    await viewer.display();
  }

  private async generateExpandedPreview(
    projectToPreview: Project,
    shouldDelete: boolean
  ) {
    const previewer = new ExpandedPreviewer(
      this.latestReport,
      this.targetId,
      projectToPreview,
      shouldDelete
    );
    await previewer.preview();
  }

  private async refreshSnapshotData() {
    this.model = await this.snapshotClient.get(this.model.id, {
      includeReports: true,
    });
  }

  public waitUntilDone(options: WaitUntilOperationDone = {}) {
    const opts = {...Snapshot.defaultWaitOptions, ...options};
    const toMilliseconds = (seconds: number) => seconds * 1e3;

    return retry(
      this.waitUntilDoneRetryFunction(opts.onRetryCb, opts.operationToWaitFor),
      // Setting the retry mechanism to follow a time-based logic instead of specifying the  number of attempts.
      {
        retries: Math.ceil(opts.wait / opts.waitInterval),
        forever: opts.wait === 0,
        minTimeout: toMilliseconds(opts.waitInterval),
        maxTimeout: toMilliseconds(opts.waitInterval),
        maxRetryTime: toMilliseconds(opts.wait),
      }
    );
  }

  private isUnsettled() {
    return Snapshot.ongoingReportStatuses.includes(this.latestReport.status);
  }

  private isGoingThroughOperation(
    operation: ResourceSnapshotsReportType
  ): boolean {
    if (this.latestReport.type === operation) {
      return true;
    }
    return false;
  }

  private waitUntilDoneRetryFunction(
    onRetryCb: (report: ResourceSnapshotsReportModel) => void,
    operationToWaitFor?: ResourceSnapshotsReportType
  ): () => Promise<void> {
    return (async () => {
      await this.refreshSnapshotData();

      const isUnsettled = this.isUnsettled();
      const isUnexpectedOperation = operationToWaitFor
        ? !this.isGoingThroughOperation(operationToWaitFor)
        : false;

      onRetryCb(this.latestReport);

      if (isUnsettled || isUnexpectedOperation) {
        throw new SnapshotOperationTimeoutError(this);
      }
    }).bind(this);
  }
}
