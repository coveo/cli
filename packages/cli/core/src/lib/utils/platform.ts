import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
