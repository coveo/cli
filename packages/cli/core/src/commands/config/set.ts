import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
