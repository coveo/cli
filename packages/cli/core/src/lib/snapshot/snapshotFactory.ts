import {
  CreateFromFileOptions,
  ResourceSnapshotsReportType,
  ResourceSnapshotType,
} from '@coveo/platform-client';
import {readFileSync} from 'fs';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {SnapshotPullModelResources} from './pullModel/interfaces';
import {Snapshot, WaitUntilDoneOptions} from './snapshot';
import {Project} from '../project/project';
import {ensureResourcesAccess, ensureSnapshotAccess} from './snapshotAccess';
import {Blob} from 'node:buffer';

export class SnapshotFactory {
  public static async createSnapshotFromProject(
    project: Project,
    targetOrg: string,
    options?: WaitUntilDoneOptions
  ): Promise<Snapshot> {
    const client = await this.getClient(targetOrg);
    await ensureResourcesAccess(client, project.resourceTypes);
    const pathToZip = await project.compressResources();
    const file = readFileSync(pathToZip);
    const blob = new Blob([file], {
      type: 'application/zip',
    });
    const computedOptions: CreateFromFileOptions = {
      developerNotes: 'cli-created-from-zip',
    };

    const model = await client.resourceSnapshot.createFromFile(
      blob,
      computedOptions
    );
    const snapshot = new Snapshot(model, client);

    await snapshot.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.CreateSnapshot,
      ...options,
    });

    return snapshot;
  }

  public static async createFromExistingSnapshot(
    snapshotId: string,
    targetOrg: string,
    options?: WaitUntilDoneOptions
  ) {
    const client = await this.getClient(targetOrg);
    await ensureSnapshotAccess(client, snapshotId);
    const model = await client.resourceSnapshot.get(snapshotId, {
      includeReports: true,
    });

    const snapshot = new Snapshot(model, client);
    await snapshot.waitUntilDone(options);

    return snapshot;
  }

  public static async createFromOrg(
    resourcesToExport: SnapshotPullModelResources,
    targetOrg: string,
    options?: WaitUntilDoneOptions
  ) {
    const client = await this.getClient(targetOrg);
    const resourceTypes = Object.keys(
      resourcesToExport
    ) as ResourceSnapshotType[];
    await ensureResourcesAccess(client, resourceTypes);
    const model = await client.resourceSnapshot.createFromOrganization(
      {resourcesToExport},
      {includeChildrenResources: true, developerNotes: 'Created by Coveo-CLI'}
    );

    const snapshot = new Snapshot(model, client);

    await snapshot.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.CreateSnapshot,
      ...options,
    });

    return snapshot;
  }

  private static async getClient(targetOrg: string) {
    return await new AuthenticatedClient().getClient({organization: targetOrg});
  }
}
