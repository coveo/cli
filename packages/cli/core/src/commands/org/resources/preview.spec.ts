jest.mock('../../../lib/decorators/preconditions/git');
jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/trackable');

jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('../../../lib/snapshot/snapshot');
jest.mock('../../../lib/snapshot/snapshotFactory');
jest.mock('../../../lib/project/project');

import {test} from '@oclif/test';
import {Project} from '../../../lib/project/project';
import {join, normalize} from 'path';
import {cwd} from 'process';
import {Config} from '@coveo/cli-commons/config/config';
import {SnapshotFactory} from '../../../lib/snapshot/snapshotFactory';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {SnapshotReporter} from '../../../lib/snapshot/snapshotReporter';
import {ResourceSnapshotsReportType} from '@coveord/platform-client';
import {
  getErrorReport,
  getMissingVaultEntryReport,
  getSuccessReport,
} from '../../../__stub__/resourceSnapshotsReportModel';
import {CliUx, Command} from '@oclif/core';

import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {IsGitInstalled} from '../../../lib/decorators/preconditions';

const mockedSnapshotFactory = jest.mocked(SnapshotFactory, true);
const mockedConfig = jest.mocked(Config);
const mockedProject = jest.mocked(Project);
const mockedConfigGet = jest.fn();
const mockedDeleteTemporaryZipFile = jest.fn();
const mockedGetResourceManifest = jest.fn();
const mockedDeleteSnapshot = jest.fn();
const mockedSaveDetailedReport = jest.fn();
const mockedAreResourcesInError = jest.fn();
const mockedValidateSnapshot = jest.fn();
const mockedPreviewSnapshot = jest.fn();
const mockedLastReport = jest.fn();
const mockedIsGitInstalled = jest.mocked(IsGitInstalled, true);
const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
const mockEvaluate = jest.fn();

const mockProject = () => {
  mockedProject.mockImplementation(
    () =>
      ({
        compressResources: () =>
          Promise.resolve(normalize(join('path', 'to', 'resources.zip'))),
        deleteTemporaryZipFile: mockedDeleteTemporaryZipFile,
        getResourceManifest: mockedGetResourceManifest,
      } as unknown as Project)
  );
};

const doMockConfig = () => {
  mockedConfigGet.mockReturnValue({
    region: 'us',
    organization: 'default-org',
    environment: 'prod',
  });

  mockedConfig.mockImplementation(
    () =>
      ({
        get: mockedConfigGet,
      } as unknown as Config)
  );
};

const mockAuthenticatedClient = () => {
  const mockGetClient = jest.fn().mockResolvedValue({
    privilegeEvaluator: {
      evaluate: mockEvaluate,
    },
  });

  mockedAuthenticatedClient.prototype.getClient = mockGetClient;
};

const mockUserHavingAllRequiredPlatformPrivileges = () => {
  mockEvaluate.mockResolvedValue({approved: true});
};

const mockUserNotHavingAllRequiredPlatformPrivileges = () => {
  mockEvaluate.mockResolvedValue({approved: false});
};

const mockSnapshotFactory = () => {
  const dummySnapshot = {
    validate: mockedValidateSnapshot,
    preview: mockedPreviewSnapshot,
    delete: mockedDeleteSnapshot,
    saveDetailedReport: mockedSaveDetailedReport,
    areResourcesInError: mockedAreResourcesInError,
    latestReport: mockedLastReport,
    id: 'banana-snapshot',
    targetId: 'potato-org',
  } as unknown as Snapshot;
  mockedSnapshotFactory.createFromZip.mockReturnValue(
    Promise.resolve(dummySnapshot)
  );
  mockedSnapshotFactory.createFromExistingSnapshot.mockReturnValue(
    Promise.resolve(dummySnapshot)
  );
};

const mockSnapshotFactoryReturningValidSnapshot = () => {
  const successReport = getSuccessReport(
    'success-report',
    ResourceSnapshotsReportType.Apply
  );
  mockedValidateSnapshot.mockImplementation(() =>
    Promise.resolve(new SnapshotReporter(successReport))
  );
  mockSnapshotFactory();
};

const mockSnapshotFactoryReturningInvalidSnapshot = () => {
  const errorReport = getErrorReport(
    'error-report',
    ResourceSnapshotsReportType.Apply
  );
  mockedValidateSnapshot.mockImplementation(() =>
    Promise.resolve(new SnapshotReporter(errorReport))
  );
  mockSnapshotFactory();
};

const mockSnapshotFactoryReturningSnapshotWithMissingVaultEntries = () => {
  const missingVaultEntry = getMissingVaultEntryReport(
    'missing-vault-entry',
    ResourceSnapshotsReportType.Apply
  );
  mockedValidateSnapshot.mockImplementation(() =>
    Promise.resolve(new SnapshotReporter(missingVaultEntry))
  );
  mockSnapshotFactory();
};

describe('org:resources:preview', () => {
  const doMockPreconditions = function () {
    const mockGit = (_target: Command) => Promise.resolve();
    mockedIsGitInstalled.mockReturnValue(mockGit);
  };

  beforeAll(() => {
    doMockConfig();
    mockProject();
    mockAuthenticatedClient();
  });

  beforeEach(() => {
    doMockPreconditions();
    mockUserHavingAllRequiredPlatformPrivileges();
  });

  afterEach(() => {
    mockEvaluate.mockClear();
    mockedIsGitInstalled.mockClear();
  });

  describe('when the report contains no resources in error', () => {
    beforeAll(() => {
      mockSnapshotFactoryReturningValidSnapshot();
    });

    afterAll(() => {
      mockedSnapshotFactory.mockReset();
    });

    test
      .do(() => {
        mockUserNotHavingAllRequiredPlatformPrivileges();
      })
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .catch(/You are not authorized to create snapshot/)
      .it('should return an error message if privileges are missing');

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('should use cwd as project', () => {
        expect(mockedProject).toHaveBeenCalledWith(cwd(), 'default-org');
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('should work with default connected org', () => {
        expect(mockedSnapshotFactory.createFromZip).toHaveBeenCalledWith(
          normalize(join('path', 'to', 'resources.zip')),
          'default-org',
          expect.objectContaining({})
        );
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview', '-o', 'myorg'])
      .it('should work with specified target org', () => {
        expect(mockedSnapshotFactory.createFromZip).toHaveBeenCalledWith(
          normalize(join('path', 'to', 'resources.zip')),
          'myorg',
          expect.objectContaining({})
        );
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview', '-s', 'some-snapshot-id'])
      .it('should work with specified snapshot id', () => {
        expect(
          mockedSnapshotFactory.createFromExistingSnapshot
        ).toHaveBeenCalledWith(
          'some-snapshot-id',
          'default-org',
          expect.objectContaining({})
        );
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('should set a 60 seconds wait', () => {
        expect(mockedSnapshotFactory.createFromZip).toHaveBeenCalledWith(
          normalize(join('path', 'to', 'resources.zip')),
          'default-org',
          {wait: 60}
        );
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview', '-w', '312'])
      .it('should set a 312 seconds wait', () => {
        expect(mockedSnapshotFactory.createFromZip).toHaveBeenCalledWith(
          normalize(join('path', 'to', 'resources.zip')),
          'default-org',
          {wait: 312}
        );
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('#validate should not take into account missing resources', () => {
        expect(mockedValidateSnapshot).toHaveBeenCalledWith(false, {wait: 60});
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview', '-d'])
      .it('#validate should take into account missing resoucres', () => {
        expect(mockedValidateSnapshot).toHaveBeenCalledWith(true, {wait: 60});
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('should preview the snapshot', () => {
        expect(mockedPreviewSnapshot).toHaveBeenCalledTimes(1);
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('should delete the compressed folder', () => {
        expect(mockedDeleteTemporaryZipFile).toHaveBeenCalledTimes(1);
      });

    test
      .stdout()
      .stderr()
      .do(() => {
        mockedValidateSnapshot.mockImplementationOnce(() => {
          throw new Error('You shall not pass');
        });
      })
      .command(['org:resources:preview'])
      .catch(() => {
        expect(mockedDeleteTemporaryZipFile).toHaveBeenCalledTimes(1);
      })
      .it('should delete the compressed folder on error');

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview', '-s', 'some-snapshot-id'])
      .it('should no delete the snapshot', () => {
        expect(mockedDeleteSnapshot).not.toHaveBeenCalled();
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .it('should display expanded preview', () => {
        expect(mockedPreviewSnapshot).toHaveBeenCalledWith(
          expect.anything(),
          expect.anything(),
          true
        );
      });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview', '--previewLevel', 'light'])
      .it('should only display light preview', () => {
        expect(mockedPreviewSnapshot).toHaveBeenCalledWith(
          expect.anything(),
          expect.anything(),
          false
        );
      });
  });
  //#region TODO: CDX-948, setup phase needs to be rewrite and assertions 'split up' (e.g. the error ain't trigger directly by the function, therefore should not be handled)
  describe('when the report contains resources in error', () => {
    beforeAll(() => {
      mockSnapshotFactoryReturningInvalidSnapshot();
    });

    beforeEach(() => {
      mockedAreResourcesInError.mockReturnValueOnce(false);
      mockedSaveDetailedReport.mockReturnValueOnce(
        normalize(join('saved', 'snapshot'))
      );
    });

    afterAll(() => {
      mockedSnapshotFactory.mockReset();
    });

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .catch(/Invalid snapshot/)
      .it('should throw an error for invalid snapshots');

    test
      .stdout()
      .stderr()
      .command(['org:resources:preview'])
      .catch((ctx) => {
        expect(ctx.message).toContain(
          'https://platform.cloud.coveo.com/admin/#potato-org/organization/resource-snapshots/banana-snapshot'
        );
      })
      .it('should print an URL to the snapshot page');
  });

  describe('when the report contains resources with missing vault entries', () => {
    beforeAll(() => {
      mockSnapshotFactoryReturningSnapshotWithMissingVaultEntries();
    });

    afterAll(() => {
      mockedSnapshotFactory.mockReset();
    });

    describe('when the user refuses to migrate or type in the missing vault entries', () => {
      test
        .stdout()
        .stderr()
        .stub(CliUx.ux, 'confirm', () => () => Promise.resolve(false))
        .command(['org:resources:preview'])
        .catch(/Your destination organization is missing vault entries/)
        .it('should throw an error for invalid snapshots');
    });
  });
  //#endregion
});
