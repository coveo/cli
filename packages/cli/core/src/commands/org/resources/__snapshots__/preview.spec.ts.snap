// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org:resources:preview when resources privileges are missing should return an error message if resource privileges are missing 1`] = `"Cannot read properties of undefined (reading 'validate')"`;

exports[`org:resources:preview when resources privileges are missing should return an error message if resource privileges are missing for specific snapshot 1`] = `
"You do not have [1mREAD[22m privilege on the snapshot some-snapshot-id
Visit https://docs.coveo.com/en/3357 for more info on privileges required to manage snapshots."
`;
